src/main/kotlin/org/rust/lang/core/psi/RsPsiTreeChangeListener.kt

```diff
@@ -0,0 +1,225 @@
+            event.propertyName,
+            event.oldValue,
+            event.newValue,
+            event.element
+        ))
+    }
+
+    override fun propertyChanged(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.PropertyChange.After(
+            event.propertyName,
+            event.oldValue,
+            event.newValue,
+            event.element
+        ))
+    }
+
+    override fun beforeChildReplacement(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildReplace.Before(
+            event.file,
+            event.parent,
+            event.oldChild,
+            event.newChild
+        ))
+    }
+
+    override fun childReplaced(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildReplace.After(
+            event.file,
+            event.parent,
+            event.oldChild,
+            event.newChild
+        ))
+    }
+
+    override fun beforeChildAddition(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildAdd.Before(
+            event.file,
+            event.parent,
+            event.child
+        ))
+    }
+
+    override fun childAdded(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildAdd.After(
+            event.file,
+            event.parent,
+            event.child
+        ))
+    }
+
+    override fun beforeChildMovement(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildMove.Before(
+            event.file,
+            event.oldParent,
+            event.newParent,
+            event.child
+        ))
+    }
+
+    override fun childMoved(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildMove.After(
+            event.file,
+            event.oldParent,
+            event.newParent,
+            event.child
+        ))
+    }
+
+    override fun beforeChildRemoval(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildRemove.Before(
+            event.file,
+            event.parent,
+            event.child
+        ))
+    }
+
+    override fun childRemoved(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildRemove.After(
+            event.file,
+            event.parent,
+            event.child
+        ))
+    }
+
+    override fun beforeChildrenChange(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildrenChange.Before(
+            event.file,
+            event.parent,
+            (event as? PsiTreeChangeEventImpl)?.isGenericChange == true
+        ))
+    }
+
+    override fun childrenChanged(event: PsiTreeChangeEvent) {
+        handleEvent(RsPsiTreeChangeEvent.ChildrenChange.After(
+            event.file,
+            event.parent,
+            (event as? PsiTreeChangeEventImpl)?.isGenericChange == true
+        ))
+    }
+}
+
+private val PsiElement.safeText
+    get() = try {
+        text
+    } catch (e: Exception) {
+        "<exception>"
```

Comment:
Hm, is exception message always useless? Is not, let's add it into returning value

Reply:
I've added a comment to explain why is it needed. If element is invalid, in general, we can't call getText(). But sometimes it works, so for debug purposes we can use it. But if not, exception message is irrelevant